packages/flutter/lib/src/rendering/selection.dart

```diff
@@ -104,17 +111,136 @@ abstract class SelectionHandler implements ValueListenable<SelectionGeometry> {
   SelectionResult dispatchSelectionEvent(SelectionEvent event);
 }
 
+/// This class stores the information of the selection under a [Selectable]
+/// or [SelectionHandler].
+///
+/// The [SelectedContentRange] for a given [Selectable] or [SelectionHandler]
+/// can be retrieved by calling [SelectionHandler.getSelection].
+@immutable
+class SelectedContentRange with Diagnosticable {
+  /// Creates a [SelectedContentRange] with the given values.
+  const SelectedContentRange({
+    required this.contentLength,
+    required this.startOffset,
+    required this.endOffset,
+  }) : assert((startOffset >= 0 && endOffset >= 0)
+              || (startOffset == -1 && endOffset == -1));
+
+  /// A selected content range that represents an empty selection, i.e. nothing
+  /// is selected.
+  const SelectedContentRange.empty({
+    int contentLength = 0,
+  }) : this(
+        contentLength: contentLength,
+        startOffset: -1,
+        endOffset: -1,
+      );
+
+  /// The length of the content in the [Selectable] or [SelectionHandler] that
+  /// created this object.
+  ///
+  /// The absolute value of the difference between the [startOffset] and [endOffset]
+  /// contained by this [SelectedContentRange] must not exceed the content length.
+  final int contentLength;
+
+  /// The start of the selection relative to the start of the content.
+  ///
+  /// {@template flutter.rendering.selection.SelectedContentRange.selectionOffsets}
+  /// For example a [Text] widget's content is in the format of an [TextSpan] tree.
+  ///
+  /// Take the [Text] widget and [TextSpan] tree below:
+  ///
+  /// {@tool snippet}
+  /// ```dart
+  /// const Text.rich(
+  ///   TextSpan(
+  ///     text: 'Hello world, ',
+  ///     children: <InlineSpan>[
+  ///       WidgetSpan(
+  ///         child: Text('how are you today? '),
+  ///       ),
+  ///       TextSpan(
+  ///         text: 'Good, thanks for asking.',
+  ///       ),
+  ///     ],
+  ///   ),
+  /// )
+  /// ```
+  /// {@end-tool}
+  ///
+  /// If we select from the beginning of 'world' to the end of the '.'
+  /// at the end of the [TextSpan] tree, the [SelectedContentRange] from
+  /// [SelectionHandler.getSelection] will be relative to the text of the
+  /// [TextSpan] tree, with [WidgetSpan] content being flattened. The [startOffset]
+  /// will be 6, and [endOffset] will be 56. This takes into account the
+  /// length of the content in the [WidgetSpan], which is 19, so the overall
+  /// [contentLength] will be 56.
+  ///
+  /// If [startOffset] and [endOffset] are both -1, the selected content range is
+  /// empty, i.e. nothing is selected.
+  /// {@endtemplate}
+  final int startOffset;
+
+  /// The end of the selection relative to the start of the content.
+  ///
+  /// {@macro flutter.rendering.selection.SelectedContentRange.selectionOffsets}
+  final int endOffset;
+
+  /// Whether this range represents a valid position in the content.
+  bool get isValid => startOffset >= 0 && endOffset >= 0;
```

Comment:
nit: shouldn't this be `isEmpty`? The constructor is called `empty` and the term "valid" is not clearly defined in the docs it seems.

Reply:
Good point, changed to `isEmpty`.